### ISSUE TITLE
Poor div rounding can lead to significant drop in numerical precision on realistic settings.

### ISSUE THREAD

**Issue by Chillee** [pytorch/pytorch#122179]
### 🐛 Describe the bug

```
import torch
torch.set_default_device('cuda')

def f(x, y):
    x = torch.matmul(x, x)
    x = x / y
    return torch.softmax(x, dim=-1)

torch.manual_seed(1234)
x = torch.tensor([[[[-16.1649,   5.6846,  -5.1022,  -9.1134],
          [-11.5552,  -2.2615, -12.8913,  10.6538],
          [ -7.1666,  -5.3333,   2.0776,  -9.7984],
          [  7.4469,  -2.3948,   2.7371,   0.9201]],

         [[ -8.0361, -16.3771,  22.7741,   4.4685],
          [ 20.8047,  -0.7771,  -2.4355,  -2.2299],
          [  3.8343,  -2.0914,  -2.4077,   2.2740],
          [-15.8663,  -2.7015, -12.5241,  -3.0040]],

         [[ -2.5139,  14.4393,  -3.7186,   1.2255],
          [  5.6742,  14.1842,  -8.5976,  16.8366],
          [ -9.7358,  -3.0279,  11.8164,  -4.0787],
          [ -9.0621,   8.2580,  29.9486,  -2.4107]],

         [[  7.3622,  12.5640, -20.5592,  13.6237],
          [-11.5640,   0.8832,  16.7275,  -2.5009],
          [ -2.0953, -12.2276, -26.2633,   4.5268],
          [ 15.3329, -11.7492,   6.5650,  -9.2483]]],


        [[[  7.9980,  -4.9369,   3.1508,   5.2994],
          [  3.8052,   3.9514,   8.4987, -10.5045],
          [ -2.6827,  -4.0010,  -4.0611,   6.4091],
          [-19.0318,   6.4073,   2.8923,   8.0250]],

         [[  7.1650,  -3.4585,   5.7720,  -5.0305],
          [ -0.9765,  -3.0086,  11.7114,   8.0555],
          [ -3.1027,  -3.5514,   9.6182,  -8.8526],
          [ -9.2348,  -6.0239,   6.2528,  -6.7221]],

         [[ 11.5936,  22.4139,  -0.4089,  -4.9889],
          [ 14.8217,  -2.3426, -17.6189,   3.7427],
          [  1.9546, -13.0902,   8.6293,  -7.2457],
          [ -7.6900,  -4.5796,   9.6332, -10.2631]],

         [[  0.8027,  -1.0955,  14.8404,  -0.2673],
          [  3.2143,  -1.8640,  -2.9678,   6.5165],
          [ -3.9865,   6.5230,   6.3019,  -0.4247],
          [  8.3185, -13.5076,  27.0986,  -1.6792]]]], device='cuda:0')
y = torch.tensor([[[ 0.6331]],

        [[ 1.6358]],

        [[-0.3459]],

        [[ 1.0196]]], device='cuda:0', requires_grad=True)
f(x, y).sum().backward()
print(y.grad)
y.grad = None
torch.compile(f)(x, y).sum().backward()
print(y.grad)
```
```
tensor([[[-8.1229e-06]],

        [[ 0.0000e+00]],

        [[ 0.0000e+00]],

        [[-2.3746e-06]]], device='cuda:0')

tensor([[[0.0203]],

        [[0.0016]],

        [[0.2055]],

        [[0.0215]]], device='cuda:0')
```

Solved by https://github.com/pytorch/pytorch/pull/115435

cc: @peterbell10 
### Versions

N/A

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire

**Comment by Chillee** [comment]
I can't repro this now....

**Comment by Chillee** [comment]
I suspect the issue here is similar to the one from attention: https://github.com/pytorch/pytorch/issues/121558#issuecomment-2008359796

**Comment by Chillee** [comment]
Here's a version of this repro that demonstrates higher numeric error.

```
import torch
torch.set_default_device('cuda')

def graph1( primals_1: "f32[2, 4, 4, 4]", primals_2: "f32[4, 1, 1]"):
    # File: /data/users/chilli/pytorch/../t.py:5 in f, code: x = x / y
    div: "f32[2, 4, 4, 4]" = torch.ops.aten.div.Tensor(primals_1, primals_2)

    # File: /data/users/chilli/pytorch/../t.py:6 in f, code: return torch.softmax(x, dim=-1)
    amax: "f32[2, 4, 4, 1]" = torch.ops.aten.amax.default(div, [-1], True)
    sub: "f32[2, 4, 4, 4]" = torch.ops.aten.sub.Tensor(div, amax);  div = amax = None
    exp: "f32[2, 4, 4, 4]" = torch.ops.aten.exp.default(sub);  sub = None
    sum_1: "f32[2, 4, 4, 1]" = torch.ops.aten.sum.dim_IntList(exp, [-1], True)
    div_1: "f32[2, 4, 4, 4]" = torch.ops.aten.div.Tensor(exp, sum_1);  exp = sum_1 = None
    return div_1

def f(x, y):
    x = x / y
    return torch.softmax(x, dim=-1)

torch.manual_seed(1234)
x = torch.tensor([[[[-16.1649,   5.6846,  -5.1022,  -9.1134],
          [-11.5552,  -2.2615, -12.8913,  10.6538],
          [ -7.1666,  -5.3333,   2.0776,  -9.7984],
          [  7.4469,  -2.3948,   2.7371,   0.9201]],

         [[ -8.0361, -16.3771,  22.7741,   4.4685],
          [ 20.8047,  -0.7771,  -2.4355,  -2.2299],
          [  3.8343,  -2.0914,  -2.4077,   2.2740],
          [-15.8663,  -2.7015, -12.5241,  -3.0040]],

         [[ -2.5139,  14.4393,  -3.7186,   1.2255],
          [  5.6742,  14.1842,  -8.5976,  16.8366],
          [ -9.7358,  -3.0279,  11.8164,  -4.0787],
          [ -9.0621,   8.2580,  29.9486,  -2.4107]],

         [[  7.3622,  12.5640, -20.5592,  13.6237],
          [-11.5640,   0.8832,  16.7275,  -2.5009],
          [ -2.0953, -12.2276, -26.2633,   4.5268],
          [ 15.3329, -11.7492,   6.5650,  -9.2483]]]
        ], device='cuda:0') * 1000

y = torch.tensor([[[ 0.6331]],

        [[ 1.6358]],

        [[-0.3459]],

        [[ 1.0196]]], device='cuda:0')

div_1 = graph1(x, y)
div_1_compile = torch.compile(graph1)(x, y)
div_1_64 = graph1(x.to(dtype=torch.float64), y.to(dtype=torch.float64))
print((div_1 - div_1_64).abs().max())
print((div_1_compile - div_1_64).abs().max())
>>> tensor(4.8773e-81, device='cuda:0', dtype=torch.float64)
>>> tensor(0.0007, device='cuda:0', dtype=torch.float64)
```

There are 3 issues that go on here (this is the code: https://pastebin.com/80pbhpLt)

1. We choose to unroll `sum/amax`, preventing fusion with the downstream `div_1`.
2. We choose to rematerialize most of the pointwise ops (`div`, `sub`, and `exp`) (i.e. we compute it in both the first kernel and the second kernel). This is also suboptimal, since we load both inputs in two kernels for no reason.
3.  In the second kernel, we are computing `((arg0_1 / arg1_1) - buf0).exp() / buf1`. Here, because `buf0` is supposed to be `(arg0_1 / arg1_1).amax()`. So, there is some element for which this should be `((arg0_1 / arg1_1) - (arg0_1 / arg1_1)).exp()`, which is equal to 1. However, because of fma, this is rewritten as `fma(arg0_1, (1/arg_1), -buf0)`, which does not equal 1.

TL;DR: fma bad.

**Comment by peterbell10** [comment]
Closing as per https://github.com/pytorch/pytorch/pull/115435#issuecomment-2166676593
